Sources/Transformer.swift

```
@@ -0,0 +1,177 @@
+//  THE SOFTWARE.
+//
+
+import Foundation
+import AlgoliaSearch
+
+/// A Transformable data source
+public protocol Transformable {
+    associatedtype IResults
+    associatedtype IParameters
+    
+    typealias SearchResultsHandler = (_ results: IResults?, _ error: Error?) -> Void
+    
+    // Search operation
+    func search(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia params to custom backend params.
+    func map(query: Query) -> IParameters
+    
+    // Transforms the custom backend result to an Algolia result.
+    func map(results: IResults) -> SearchResults
+    
+    
+    // Disjunctive Search operation
+    func searchDisjunctiveFaceting(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia disjunctive params to custom backend params.
+    func map(query: Query, disjunctiveFacets: [String], refinements: [String : [String]]) -> IParameters
+    
+    
+    // Search for facet value operation
+    func searchForFacetValues(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia facet value params to custom backend params.
+    func map(query: Query?, facetName: String, matching text: String) -> IParameters
+}
+
+/// Base class used to implement a custom backend with Algolia
+/// You need to make sure to implement the following method:
+/// - search(query:searchResultsHandler:)
+/// - map(query:)
+/// - map(results:)
+open class SearchTransformer<Parameters, Results>: Transformable, Searchable {
+    
+    public typealias IParameters = Parameters
+    public typealias IResults = Results
+    
+    public init() {}
+    
+    // Transformable protocol
+    
+    open func search(_ query: Parameters, searchResultsHandler: @escaping SearchResultsHandler) {
+        fatalError("make sure to override search(query:searchResultsHandler:) for custom backend")
```

This `fatalError` is only raised if the `search()` method is not implemented, right?

---

Exactly: when someone tries to implement the `SearchTransformer` class, if he forget this method, he'll automatically get a crash. This is to force him to implement the method. The reason why it's done this way (which is a bit dirty) is due to Swift limitations that I described in the PR description

---

Just wanted to make sure! I decided to return empty results (fail silently) but to warn in the console. Happy to go either way though.

I think this is fine to do like this in Swift.

---

On second thought, you're right. I now throw errors if the method is not implemented.